xonsh/aliases.py

```diff
@@ -94,6 +95,10 @@ def source_foreign(args, stdin=None):
         if k in denv and v == denv[k]:
             continue  # no change from original
         env[k] = v
+    # If run in un-safe mode we are sure the command completed correctly,
+    # thus we can remove any env-vars that were unset by the script.
+    if not ns.safe:
+        [env.pop(k, None) for k in denv if k not in fsenv]
```

Comment:
Ahh this is an interesting point that I hadn't considered.  Thanks!  Would you mind turning this into a simple loop though? There is no reason to make a list of a bunch of values that will be discarded.